src/lib.rs

```diff
@@ -1472,28 +1473,52 @@ pub mod __rt {
         }
 
         pub fn borrow(&self) -> Ref<T> {
-            unsafe {
-                if self.borrow.get() == usize::max_value() {
-                    borrow_fail();
-                }
-                self.borrow.set(self.borrow.get() + 1);
-                Ref {
-                    value: &*self.value.get(),
-                    borrow: &self.borrow,
+            let mut borrow = self.borrow.load(Ordering::SeqCst);
+            loop {
+                if borrow == usize::max_value() {
+                    borrow_fail()
                 }
+                match self.borrow.compare_exchange_weak(
+                    borrow,
+                    borrow + 1,
+                    Ordering::SeqCst,
+                    Ordering::SeqCst,
+                ) {
+                    Ok(_) => unsafe {
+                        return Ref {
+                            value: &*self.value.get(),
+                            borrow: &self.borrow,
+                        };
+                    },
+                    Err(new_borrow) => {
+                        borrow = new_borrow;
+                    }
+                };
             }
```

Comment:
I'm assuming the `loop` and `compare_exchange()` is necessary because you want to make sure that your overflow check is actually correct.

But I would think that using `fetch_add()` should be good enough, because it returns the previous value, which then can be checked instead.

Reply:
> But I would think that using `fetch_add()` should be good enough, because it returns the previous value, which then can be checked instead.

What if someone else sets `self.borrow.store(usize::max_value())` before we call `self.borrow.fetch_add(1)`? I think it would overflow to `0` and unlock the value! Sure we could check the previous value, but untold amounts of damage could happen while we're doing so.